### DISCUSSION TITLE
parallelization expression graph and jacobian

### DISCUSSION THREAD

**Discussion by charleskhazoom** [casadi/casadi#3197] [Q&A] [answered]
When building the constraints with a multiple-shooting method, most of constraint computation can be parallelized across each shooting node (e.g. using the `map` functionality). So the jacobian of the corresponding constraints can also be parallelized. My question is : if I used `map` to compute the constraints in parallel, will the corresponding Jacobian computation also be parallelized ? From my testing it does not seem to be the case. Is there a way to achieve this automatically, besides building the jacobian manually?

**Comment by jgillis** [comment]
This is supposed to work. I'll have a look..

**Reply by charleskhazoom** [reply]
Also, note that I am using `expand=true`. Is is possible that the parallelization would only works with MX, or should it work with SX too ?

**Reply by charleskhazoom** [reply]
I just checked, and it seems that when I am using Ipopt from the opti stack and let Casadi form the functions for derivatives, the computation of the derivates is indeed parallelized, as evidenced by the stats (e.g  `t_wall_nlp_jac_g: 0.3017` vs  `t_proc_nlp_jac_g: 4.7497`). This is the case whether I use MX or SX.

However, I need to form `J = jacobian(opti.g, opti.x)` manually, and pass it to another qp solver not interfaced by Casadi. Only in this case the computation of `J` is not parallelized even if I use `map(N, 'thread',10)` to form `opti.g`. How to parallelize the computation of `J` in this type of use case?

**Reply by jaeandersson** [reply, marked as answer]
Correct, `expand=true` will in general destroy parallelization. There are ways around that, by defining a custom Jacobian function for example, by that's a bit involved. My suggestion would be not to do `expand=true`, or if you do it, just do it on a lower-level function.
